caffe2/core/dispatch/DispatchTable.h

```diff
@@ -0,0 +1,150 @@
+    // which we need.  It's worth spelling out exactly what we need:
+    //
+    //  - We need LOCK FREE read access to the table (as per the performance benchmark
+    //    at https://fb.quip.com/hvz3AGnx8MQ8
+    //
+    //  - We need to support writes which are possibly concurrent with reads, occurring when
+    //    a dynamic library is loaded or unloaded.
+    //
+    //  - We probably can require that dynamic library loads/unloads be synchronized (so
+    //    there are never two concurrent loads.)
+
+    template<class Key_>
+    void emplace(Key_&& key, void* value) {
+      // TODO Locking
+      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+      auto result = map_.emplace(std::forward<Key>(key), value);
+      if (!result.second) {
+        throw std::logic_error("Tried to register conflicting operators to the dispatcher.");
+      }
+    }
+
+    void erase(const Key& key) {
+      // TODO Locking
+      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+      size_t num_removed = map_.erase(key);
+      assert(num_removed <= 1); //This is not a multi-map
+      if (num_removed == 0) {
+        throw std::logic_error("Tried to deregister an operator that isn't registered.");
+      }
+    }
+
+    void* lookup(const Key& key) const {
+      // TODO (lock needed but slow perf. Find better way)
+      // std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+      auto found = map_.find(key);
+      if (found == map_.end()) {
+        return nullptr;
+      } else {
+        return found->second;
+      }
+    }
+
+private:
+    ska::flat_hash_map<Key, void*> map_;
+    // TODO Figure out how to get fast locking in C++11 (use boost::shared_timed_mutex? folly::SharedMutex?)
+    //mutable std::shared_timed_mutex mutex_;
+};
+} // namespace details
+
+/**
+ * Per-operator dispatch table.
+ *
+ * Given an operator specified by 'OpSchemaDef', this class records a dispatch table for
+ * various backends provided for this operator.  For example, if we consider the operator
+ * add(Tensor, Tensor), the dispatch table for this operator may contain implementations
+ * for various dynamic tensor types, such as (CPUFloatTensor, CPUFloatTensor),
+ * (CUDAFloatTensor, CUDAFloatTensor), etc.
+ *
+ * @tparam OpSchemaDef The operator signature this dispatch table encodes.
+ */
+// TODO: Support dispatch for meta-operators (which apply to all dynamic types)
+template<class OpSchemaDef>
+class DispatchTable final {
+private:
+  using Schema = OpSchema<OpSchemaDef>;
+
+public:
+  DispatchTable(): ops_() {}
+
+  /**
+   * Register an operator implementation in the table at some dispatch key.
+   * @param func Concrete function implementation to register
+   * @param dispatch_key Dispatch key to implement this function with
+   */
+  void registerOp(typename Schema::signature::func_type* func, typename Schema::dispatch::dispatch_key_type dispatch_key) {
```

Comment:
Getting terminology straight - if the "op" means particular "name", we probably should call this one Kernel or OpImplementation

Reply:
True. I also fixed it in some other places.